### PR TITLE
Code to retrive unknown notification and UI

### LINKED ISSUE #2298: Adding a feature to dismiss already read notifications
**Shall we add a feature to dismiss your old notifications in the NotificationsActivity?** 

- That may be added via a dismiss button or by swiping the item from notification list
-  In case of zero notification we can display "No recent notification"
- Also in the notification icon( that appears in the toolbar in the main activity) that currently shows a red circle in case of notification  we can add a text view to show the number of unread notifications.

Would you like to work on this issue?
Yes

### PR DESCRIPTION
**Pull request by vanshikaarora**
**Notification Activity now fetches only unread notifications nd has UI implementation for marking notification as read on swipe**

Fixes #2298 Adding a feature to dismiss already read notifications

What changes did you make and why?

- Modified Notification Activity for adding swipe features
- API request modified to receive only unread notifications
- Added API requests for marking notification as read
- Modified Notification model class for notification Id

### CODE REVIEW

**Comment by maskaravivek** [comment]
@vanshikaarora Thanks for the PR. I have raised a PR on your fork to make the API work. If it works, you can simply merge it to automatically update this PR with that commit.

**Comment by vanshikaarora** [comment]
> * showing a dialog seems to be obstructive. On swiping left, we can show an option to mark as read. You find examples in apps like whatsapp/Gmail

@maskaravivek Do you mean showing a coloured Image like archieve in Gmail then Showing a Toast to undo that?

File: build.gradle [hunk 1]
```diff
@@ -7,7 +7,7 @@ buildscript {
         google()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.1.4'
+        classpath 'com.android.tools.build:gradle:3.2.0'
```
[the remark this answers is not in the document]
Comment (vanshikaarora) [reply build.gradle]: I'll do that

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.6-all.zip
```
[the remark this answers is not in the document]
Comment (vanshikaarora) [reply gradle/wrapper/gradle-wrapper.properties]: I'll make the changes :+1:

**Comment by maskaravivek** [comment]
> > * showing a dialog seems to be obstructive. On swiping left, we can show an option to mark as read. You find examples in apps like whatsapp/Gmail
> 
> @maskaravivek Do you mean showing a coloured Image like archieve in Gmail then Showing a Toast to undo that?

Yup, IMO that would provide a better user experience. Am fine merging this PR and enhancing the experience in a separate PR. 

Can you retest the PR once and ensure that the basic functionality is working(for the changes you did).

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java [hunk 3]
```diff
@@ -282,7 +282,9 @@ public boolean onCreateOptionsMenu(Menu menu) {
             // TODO: used vectors are not compatible with API 19 and below, change them
             menu.findItem(R.id.notifications).setIcon(ContextCompat.getDrawable(this, R.drawable.ic_notification_white_clip_art));
         } else {
-            menu.findItem(R.id.notifications).setIcon(ContextCompat.getDrawable(this, R.drawable.ic_notification_white_clip_art_dot));
+            final View notification=menu.findItem(R.id.notifications).getActionView();
+            txtViewCount=notification.findViewById(R.id.notification_count_badge);
+            txtViewCount.setText(String.valueOf(1));
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: is this value hardcoded?

**Comment by vanshikaarora** [comment]
>Can you retest the PR once and ensure that the basic functionality is working(for the changes you did).

@maskaravivek  There is some error here It returns an empty string even for a valid notificationId was it the same for you?
String result = api.action("echomarkread")
                .param("token", getEditToken())
                .param("list", notification.notificationId)
                .post()
                .getString("/api/query/echomarkread/@result");

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java (hunk 3 above)
Comment (vanshikaarora) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: Yes presently, Because I didn't had the number of unread notifications

**Comment by maskaravivek** [comment]
Can you check the full XML response in the logs.

**Comment by vanshikaarora** [comment]
>Can you check the full XML response in the logs.

@maskaravivek Can you please elaborate, which XML response you are talking about?

**Comment by maskaravivek** [comment]
@vanshikaarora if you check logcat you can see log for the API response. Similar to 

```
<?xml version="1.0"?>
<api>
  <query>
    <echomarkread result="success" rawcount="9" count="9">
      <alert rawcount="3" count="3" />
      <message rawcount="6" count="6" />
    </echomarkread>
  </query>
</api>
```

Instead of the success response, you might be getting some error.

**Comment by vanshikaarora** [comment]
@maskaravivek This was the response in my case:
 API response for method https://commons.wikimedia.beta.wmflabs.org/w/api.php is

   `  <?xml version="1.0" encoding="UTF-8"?>
           <api servedby="deployment-mediawiki-09">
                <error code="badtoken" info="Invalid CSRF token." xml:space="preserve">
See https://commons.wikimedia.beta.wmflabs.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &amp;lt;https://lists.wikimedia.org/mailman/listinfo/mediawiki-api-announce&amp;gt; for notice of API deprecations and breaking changes.</error>
</api>
`
I am receiving error due to `Invalid CSRF Token` Was it the same for you?

**Comment by maskaravivek** [comment]
Hmm, the error is `Invalid CSRF token`. For me the API request was successful. 

Try logging out and logging in. We have some authentication issues(session cookie handling related) in our app.

**Comment by vanshikaarora** [comment]
I tried re-login and then faced the same issue.
If this issue is due to my Invalid CSRF token and works perfectly for you then shall I move forward with other tasks to be completed for this PR?

**Comment by maskaravivek** [comment]
Yes, you can move ahead to other tasks to bring this PR to completion. 

I had tested it on my branch. I will retest this tonight and see if I get the same error.

**Comment by vanshikaarora** [comment]
ok I'll complete the other tasks as well

**Comment by vanshikaarora** [comment]
> showing a dialog seems to be obstructive. On swiping left, we can show an option to mark as read. You find examples in apps like whatsapp/Gmail

@maskaravivek Shall I use [this](https://github.com/daimajia/AndroidSwipeLayout)?

**Comment by maskaravivek** [comment]
The library looks good to me. Go ahead and use it.

**Comment by maskaravivek** [comment]
> @maskaravivek I have implemented the swipe to delete feature

Awesome! Can you update the PR with it?

> the number of unread notifications is currently maintained in the application. Or For getting the notification count should I make an API request to fetch number of notifications?

IMO, we should make an API call and then also cache the value. Next time when the user opens the app, we can first display the cached value and then asynchronously load the fresh value.

**Comment by vanshikaarora** [comment]
I have sent the PR for the work completed as of now
Kindly review and share your feedback's :)

**Comment by vanshikaarora** [comment]
Here is the demo of how the notification activity works currently
ezgif com-video-to-gif 3

**Comment by maskaravivek** [comment]
@vanshikaarora The swipe layout has just red background. I was hoping to have an image icon similar to one shown in the library's demo. https://github.com/daimajia/AndroidSwipeLayout

**Comment by vanshikaarora** [comment]
>Oh, I thought you are using this library. In case you are not using any library then simply add a "Mark as read" label on the red background.

@maskaraviek if need to implement this then I have to use a third party library that might conflict with the present gradle

 For the swipe feature I just came across another method

- Show delete icon on swiping item
- Then show a snackbar with an option to UNDO

Any suggestions?

**Comment by vanshikaarora** [comment]
>IMO, we should make an API call and then also cache the value. Next time when the user opens the app, we can first display the cached value and then asynchronously load the fresh value.

Shall I store it in shared preference?

**Comment by maskaravivek** [comment]
> For the swipe feature I just came across another method

> Show delete icon on swiping item
> Then show a snackbar with an option to UNDO

Showing delete icon on swiping would be nice. IMO, undo is not needed. 

Also, the API call is working fine for me. Is it still not working for you? 

> Shall I store it in shared preference?

Yes, but make sure that you manage the shared prefs properly else it might become a source of bugs. 

In its current state, I am happy to merge the PR. Can you open a separate issue for the notification count?

**Comment by vanshikaarora** [comment]
>Also, the API call is working fine for me. Is it still not working for you?

No the API call isn't working for me

>In its current state, I am happy to merge the PR. Can you open a separate issue for the notification count?

Yes sure, I'll create a separate issue for notification count

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java [hunk 4]
```diff
@@ -282,7 +290,9 @@ public boolean onCreateOptionsMenu(Menu menu) {
             // TODO: used vectors are not compatible with API 19 and below, change them
             menu.findItem(R.id.notifications).setIcon(ContextCompat.getDrawable(this, R.drawable.ic_notification_white_clip_art));
         } else {
-            menu.findItem(R.id.notifications).setIcon(ContextCompat.getDrawable(this, R.drawable.ic_notification_white_clip_art_dot));
+            final View notification=menu.findItem(R.id.notifications).getActionView();
+            txtViewCount=notification.findViewById(R.id.notification_count_badge);
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: Use butterknife. 

https://github.com/commons-app/apps-android-commons/wiki/Code-style#butterknife

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java [hunk 5]
```diff
@@ -292,6 +302,21 @@ public boolean onCreateOptionsMenu(Menu menu) {
         return true;
     }
 
+    private int getNotificationCount() {
+        final int[] size = new int[1];
+        Observable.fromCallable(() -> notificationController.getNotifications())
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(notificationList -> {
+                    Collections.reverse(notificationList);
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: Do `setText` here.

```
txtViewCount.setText(String.valueOf(notificationList.size()));
```

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java (hunk 4 above)
Comment (vanshikaarora) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: Does that means declaring it globally like this
```
@BindView(R.id.notifications)
    MenuItem menuItem;

```

Comment (maskaravivek) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: Sorry, I didn't notice that `txtViewCount` is part of `notification` view. Ignore my comment.

Comment (vanshikaarora) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: OK I'll revert the change

**Comment by maskaravivek** [comment]
Where does the notification count actually appear? Can you post a screenshot? I see just a red dot without any number on it.

**Comment by maskaravivek** [comment]
Am not sure why its null. Will check tonight.

**Comment by vanshikaarora** [comment]
>Am not sure why its null. Will check tonight.

Thanks @maskaravivek  Do have a look whenever you have time :)

**Comment by maskaravivek** [comment]
Also, nothing happens on clicking on the notification icon. Notification activity should open?

**Comment by vanshikaarora** [comment]
>Also, nothing happens on clicking on the notification icon. Notification activity should open?It was

Yes that opening the notification activity there is some issue with the current version there is no call to `onOptionsItemSelected` on clicking the notification activity

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java [hunk 6]
```diff
@@ -286,33 +287,32 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
-                    Timber.e(throwable, "Error occurred while loading notifications");
-                });
+                .subscribe(this::initNotificationViews,
+                        throwable -> Timber.e(throwable, "Error occurred while loading notifications"));
+    }
+
+    private void initNotificationViews(List<Notification> notificationList) {
+        Timber.d("Number of notifications is %d", notificationList.size());
+        final View notification = notificationsMenuItem.getActionView();
+        TextView notificationCount = notification.findViewById(R.id.notification_count_badge);
```
Comment (vanshikaarora) [inline app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java]: @maskaravivek I am still getting a NULL pointer exception here could you please help :)

**Comment by vanshikaarora** [comment]
@maskaravivek I have made the following changes:
>Also, nothing happens on clicking on the notification icon. Notification activity should open?

- Notification activity opens on tapping the icon


>the notification count background is oval. can you make it circular?

- Made it circular

>notification activity shows empty notifications. I guess it could be because of the swipe layout you added.

- Yes that was because of the Swipe Layout modified the code accordingly


Empty notification activity is now differently visible like:-
whatsapp image 2019-02-02 at 12 26 15 am

### Issues remaining now
I am still facing that NULL pointer exception:(
In notification` activity->removeNotification()` `controller.markAsRead(notification) ` gives a `NULL POINTER` exception. Can you please look into this.
ezgif com-video-to-gif 5

**Comment by maskaravivek** [comment]
@vanshikaarora Thanks for the update. Am looking into the NPE. Can you change the "Delete" icon to a "Tick" icon? The user is marking the notification as read.

**Comment by vanshikaarora** [comment]
>@vanshikaarora Thanks for the update. Am looking into the NPE. Can you change the "Delete" icon to a "Tick" icon? The user is marking the notification as read.

I'll update that :+1:

**Comment by maskaravivek** [comment]
@vanshikaarora I have raised a PR to fix other pending issues with this PR. It would be great if you could test it once after merging and update here. :)

**Comment by vanshikaarora** [comment]
>@vanshikaarora I have raised a PR to fix other pending issues with this PR. It would be great if you could test it once after merging and update here. :)

OK I'll do the :+1:

**Comment by vanshikaarora** [comment]
@maskaravivek I have made all the necessary changes and tested the PR.

The features that were expected are fully implemented I hope the PR is now ready to be merged :)
Kindly review and inform if you want any other changes in the current state :)

File: app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java [hunk 7]
```diff
@@ -587,6 +588,7 @@ public String revisionsByFilename(String filename) throws IOException {
                     .param("meta", "notifications")
                     .param("notformat", "model")
                     .param("notwikis", "wikidatawiki|commonswiki|enwiki")
+                    //.param("notfilter","!read")
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java]: uncomment

Comment (vanshikaarora) [reply app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java]: I am sorry for that I did that for testing

**Comment by vanshikaarora** [comment]
>Works well for me! Great work @vanshikaarora for picking up a decently big task and bringing it to completion. I will merge it once Travis build completes.

Thanks @maskaravivek I would always love to work on such challenging tasks :)

**Comment by vanshikaarora** [comment]
Thanks for approving the PR :)

**Pull request merged**